### DISCUSSION TITLE
Implementing the bosonic t-J model in QuSpin

### DISCUSSION THREAD

**Discussion by timharris977** [QuSpin/QuSpin#591] [Q&A] [answered]
Hi all, 

I'm looking for some advice on how to implement the **_bosonic t-J model_** with QuSpin. The model has the following Hamiltonian:

$\displaystyle{H_{t-J} = -t \sum_{\langle i,j \rangle \sigma} \left(a^\dagger_{i\sigma} a_{j\sigma} + h.c. \right)+ J\sum_{\langle i,j \rangle}\left(\vec{S}_i \cdot \vec{S}_j +  \frac{3}{4} n_i n_j\right)}$, with $\sigma \in \{\uparrow,\downarrow\}$

where $\hat{a}^{(\dagger)}_{i\sigma}$ are the bosonic spin-1/2 annihilation (creation) operators, and $\vec{S}_i = (S^x_i,S^y_i,S^z_i)$ is the tuple of local spin-1/2 operators. 

There is also a constraint on the local basis - consisting of the states: $\ket{0}, \ket{\uparrow}, \ket{\downarrow}$ - so that there can be no double occupation (i.e. it is not possible to have two bosons in different spin states on one site $\ket{\uparrow \downarrow}$).

If possible, I would also like to implement this model with $U(1) \times U(1)$ symmetry. Corresponding to conserved total particle number $\hat{N} = \sum_i \hat{n}_i$ and global magnetisation $\hat{S}_z = \sum_i \hat{S}^z_i$. 

I am quite experinced using QuSpin. However, I'm not sure if it would it be possible to implement this Hamiltonian with QuSpin's user-defined basis for example? Or if there an easier way?

Any advice would be much appreciated. Thanks!

**Comment by PFmix** [comment]
Hi,

My suggestion is late, but here is how I would do it *not* using `user_basis` (that I also haven't understood from the documentation):

- The implementation of the model is relatively standard using `spinful_fermion_basis_1d`, except for the $S$ operators. For those, I would rewrite the operators in terms of the $\hat{a}$ operators.
- For the constraint: in general, it is a bit of a nightmare without `user_basis`... lucky for you, your exact constraint has an explicit example provided [here](https://quspin.github.io/QuSpin/examples/example13.html#example13-label) that you can copy-paste. It uses an option in `spinful_fermion_basis_1d` called `double_occupancy=False`.
- The option `Nf=(int1,int2)` in `spinful_fermion_basis_1d` fixes the number of fermion of spin up and spin down, so you can fix the particle number and global magnetization sector. You then write the Hamiltonian exclusively for these values strongly implementing the symmetries you wish. If you generate several basis for various `Nf`, and their matching Hamiltonian, you effectively write and treat the Hamiltonian in its diagonal form. It will also make most computations much faster.

**Reply by timharris977** [reply]
Hi, 

Thanks for your suggestion! I agree that this approach would work to implement the standard t-J model with spin-1/2 fermions. However, here I'm specifically interested in the _**bosonic t-J model**_, where the creation and annihilation operators satisfy the commutation relations

$$ [a_{i\sigma},a^\dagger_{j\sigma'}]=\delta_{ij}\delta_{\sigma\sigma'}, \quad [a_{i\sigma},a_{j\sigma'}]= [a^\dagger_{i\sigma},a^\dagger_{j\sigma'}] = 0.$$

In this case, I think the `spinful_fermion_basis_1d` would give incorrect particle statistics. I'm also primarily interested in implementing this model for a 2D cylindrical lattice geometry (i.e. open BCs along the $x$-axis, and periodic boundaries along the $y$-axis). Could this model could be implemented with the `user_basis` instead?

**Reply by PFmix** [reply, marked as answer]
Indeed, I had missed the "bosonic" part, my apologies.
I suppose that `user_basis` is the only way to go then.
